ILSpy.ReadyToRun/ReadyToRunLanguage.cs

```diff
@@ -103,15 +103,52 @@ public override void WriteCommentLine(ITextOutput output, string comment)
 			output.WriteLine("; " + comment);
 		}
 
+		private Dictionary<ulong, UnwindCode> WriteUnwindInfo(ReadyToRunMethod readyToRunMethod, ITextOutput output)
+		{
+			IReadOnlyList<RuntimeFunction> runTimeList = readyToRunMethod.RuntimeFunctions;
+			Dictionary<ulong, UnwindCode> unwindCodes = new Dictionary<ulong, UnwindCode>();
+			foreach (RuntimeFunction i in runTimeList) {
+				if (i.UnwindInfo is UnwindInfo amd64UnwindInfo) {
+					string parsedFlags = "";
+					if ((amd64UnwindInfo.Flags & (int)UnwindFlags.UNW_FLAG_EHANDLER) != 0) {
+						parsedFlags += " EHANDLER";
+					}
+					if ((amd64UnwindInfo.Flags & (int)UnwindFlags.UNW_FLAG_UHANDLER) != 0) {
+						parsedFlags += " UHANDLER";
+					}
+					if ((amd64UnwindInfo.Flags & (int)UnwindFlags.UNW_FLAG_CHAININFO) != 0) {
+						parsedFlags += " CHAININFO";
+					}
+					if (parsedFlags.Length == 0) {
+						parsedFlags = " NHANDLER";
+					}
+					WriteCommentLine(output, $"UnwindInfo:");
+					WriteCommentLine(output, $"Version:            {amd64UnwindInfo.Version}");
+					WriteCommentLine(output, $"Flags:              0x{amd64UnwindInfo.Flags:X2}{parsedFlags}");
+					WriteCommentLine(output, $"FrameRegister:      {((amd64UnwindInfo.FrameRegister == 0) ? "none" : amd64UnwindInfo.FrameRegister.ToString())}");
+					for (int unwindCodeIndex = 0; unwindCodeIndex < amd64UnwindInfo.CountOfUnwindCodes; unwindCodeIndex++) {
+						unwindCodes.Add(amd64UnwindInfo.UnwindCodeArray[unwindCodeIndex].CodeOffset, amd64UnwindInfo.UnwindCodeArray[unwindCodeIndex]);
```

Comment:
The `CodeOffset` value here corresponds to its offset from the beginning of the `i` (i.e. The `RuntimeFunction` object). The name `RuntimeFunction` is really unfortunate, we cannot rename it as it is part of the Windows API, but it doesn't really mean what you think it is. In particular, it is not the whole method. A `ReadyToRunMethod` could have multiple `RuntimeRunction` objects that correspond to different portions of the method's code that has its own unwind data.

In practice, this is used for funclets. A funclet is a runtime construct used to support exception filtering and exception handling. A when block or a catch block could potentially access the local variables of the method, therefore a catch block is not an independent function. An implementation trick to accomplish this is to set the frame pointer to be the same the parent frame and access the local variables that way, but keep the stack pointer unchanged (in case of filter) to make sure the original stack stay intact or set the stack pointer to the frame's stack pointer (in case of catch) to represent a complete unwind. In both case, and exception or a GC could happen while the filter/catch is in progress and therefore we still need to be able to unwind there, and we will need special provision to deal with such tricky frames. This is why funclets are their own `RuntimeFunction`.

The code here is buggy because if we happen to have two runtime functions, and both of them have an unwind opcode at code offset 0, we will be adding duplicated dictionary entries.

Reply:
I believe I have fixed this by calling WriteUnwindInfo once per RunTimeFunction instead of once per ReadyToRunMethod, and so a code offset 0 for one RunTimeFunction will not overlap with the code offset 0 of another RunTimeFunction.